streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -1522,6 +1525,92 @@ private void maybeScheduleFollowupRebalance(final long encodedNextScheduledRebal
+                    return AssignmentError.ACTIVE_TASK_ASSIGNED_MULTIPLE_TIMES;
+                }
+                activeTasks.add(task.id());
+            }
+        }
+
+        // Check for ACTIVE_AND_STANDBY_TASK_ASSIGNED_TO_SAME_KAFKASTREAMS
+        for (final KafkaStreamsAssignment assignment : assignments) {
+            final Set<TaskId> activeTasksForAssignment = new HashSet<>();
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.ACTIVE) {
+                    activeTasksForAssignment.add(task.id());
+                }
+            }
+
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.STANDBY) {
+                    if (activeTasksForAssignment.contains(task.id())) {
+                        return AssignmentError.ACTIVE_AND_STANDBY_TASK_ASSIGNED_TO_SAME_KAFKASTREAMS;
+                    }
+                }
+            }
+        }
+
+        // Check for INVALID_STANDBY_TASK
+        final Set<TaskId> standbyTasksInOutput = new HashSet<>();
+        for (final KafkaStreamsAssignment assignment : assignments) {
+            for (final KafkaStreamsAssignment.AssignedTask task : assignment.assignment()) {
+                if (task.type() == KafkaStreamsAssignment.AssignedTask.Type.STANDBY) {
+                    standbyTasksInOutput.add(task.id());
+                }
+            }
+        }
+        for (final TaskInfo task : applicationState.allTasks()) {
+            if (!task.isStateful() && standbyTasksInOutput.contains(task.id())) {
+                return AssignmentError.INVALID_STANDBY_TASK;
+            }
+        }
+
+        // Check for MISSING_PROCESS_ID
+        final Map<ProcessId, KafkaStreamsState> clientStates = applicationState.kafkaStreamsStates(false);
+        final Set<ProcessId> clientsInOutput = assignments.stream().map(KafkaStreamsAssignment::processId)
+            .collect(Collectors.toSet());
+        for (final Map.Entry<ProcessId, KafkaStreamsState> entry : clientStates.entrySet()) {
+            final ProcessId processIdInInput = entry.getKey();
+            if (!clientsInOutput.contains(processIdInInput)) {
+                return AssignmentError.MISSING_PROCESS_ID;
+            }
+        }
+
+        // Check for UNKNOWN_PROCESS_ID
+        final Set<ProcessId> clientsInInput = clientStates.entrySet().stream().map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+        for (final ProcessId processIdInOutput : clientsInOutput) {
+            if (!clientsInInput.contains(processIdInOutput)) {
+                return AssignmentError.UNKNOWN_PROCESS_ID;
+            }
+        }
+
+        // Check for UNKNOWN_TASK_ID
+        final Set<TaskId> taskIdsInInput = applicationState.allTasks().stream().map(TaskInfo::id)
```

Comment:
Ah, changing the `ApplicationState#allTasks` method to return a `Map<TaskId, TaskInfo>` map would let us simplify this as well. Let's make that change, I'll update the KIP.

Then we can remove this line altogether and move the `if (!applicationState.allTasks().containsKey(task.id())) --> return AssignmentError.UNKNOWN_TASK_ID;` inside the single loop I outlined in a comment above